interfaces/prompting/requestprompts/requestprompts_test.go

```diff
@@ -1070,3 +1081,299 @@ func (s *requestpromptsSuite) TestPromptMarshalJSON(c *C) {
 
 	c.Assert(string(marshalled), Equals, string(expectedJSON))
 }
+
+func (s *requestpromptsSuite) TestPromptExpiration(c *C) {
+	var timer *testtime.TestTimer
+	restore := requestprompts.MockTimeAfterFunc(func(d time.Duration, f func()) timeutil.Timer {
+		if timer != nil {
+			c.Fatalf("created more than one timer")
+		}
+		timer = testtime.AfterFunc(d, f)
+		return timer
+	})
+	defer restore()
+
+	replyChan := make(chan notify.FilePermission, 1)
+	restore = requestprompts.MockSendReply(func(listenerReq *listener.Request, allowedPermission any) error {
+		allowedFilePermission, ok := allowedPermission.(notify.FilePermission)
+		c.Assert(ok, Equals, true)
+		replyChan <- allowedFilePermission
+		return nil
+	})
+	defer restore()
+
+	metadata := &prompting.Metadata{
+		User:      s.defaultUser,
+		Snap:      "firefox",
+		Interface: "home",
+	}
+	path := "/home/test/foo"
+	requestedPermissions := []string{"read", "write", "execute"}
+	remainingPermissions := []string{"write", "execute"}
+
+	noticeChan := make(chan noticeInfo, 1)
+	pdb, err := requestprompts.New(func(userID uint32, promptID prompting.IDType, data map[string]string) error {
+		c.Assert(userID, Equals, s.defaultUser)
+		noticeChan <- noticeInfo{
+			promptID: promptID,
+			data:     data,
+		}
+		return nil
+	})
+	c.Assert(err, IsNil)
+	defer pdb.Close()
+
+	// Add prompt
+	listenerReq := &listener.Request{}
+	prompt, merged, err := pdb.AddOrMerge(metadata, path, requestedPermissions, remainingPermissions, listenerReq)
+	c.Assert(err, IsNil)
+	c.Assert(merged, Equals, false)
+	checkCurrentNotices(c, noticeChan, prompt.ID, nil)
+
+	// Check that prompt has not immediately expired
+	c.Assert(timer.FireCount(), Equals, 0)
+
+	// Prompt should *not* expire after half of initialTimeout
+	timer.Elapse(requestprompts.InitialTimeout / 2)
+	c.Assert(timer.FireCount(), Equals, 0)
+
+	// Add another prompt, check that it does not bump the activity timeout
+	listenerReq = &listener.Request{}
+	otherPath := "/home/test/bar"
+	prompt2, merged, err := pdb.AddOrMerge(metadata, otherPath, requestedPermissions, remainingPermissions, listenerReq)
+	c.Assert(err, IsNil)
+	c.Assert(merged, Equals, false)
+	checkCurrentNotices(c, noticeChan, prompt2.ID, nil)
+
+	// Prompt should expire after initialTimeout, but half already elapsed
+	timer.Elapse(requestprompts.InitialTimeout / 2)
```

Comment:
Nitpick on arithmetic. I would elapse the remaining time, computed as
```go
remaining := InitialTimeout - initialTimeout / 2
```

This should be safe for anything that is not nicely divisible by two.